File: src/main/java/com/robohorse/robopojogenerator/generator/postprocessing/PostProcessorFactory.java
package com.robohorse.robopojogenerator.generator.postprocessing;

import com.robohorse.robopojogenerator.generator.consts.annotations.AnnotationEnum;
import com.robohorse.robopojogenerator.injections.Injector;
import com.robohorse.robopojogenerator.models.GenerationModel;

import javax.inject.Inject;

/**
 * Created by vadim on 23.10.16.
 */
public class PostProcessorFactory {
    @Inject
    public PostProcessorFactory() {
    }

    public BasePostProcessor createPostProcessor(GenerationModel generationModel) {
        if (generationModel.isUseKotlin()) {
            return Injector.getAppComponent().newKotlinDataClassPostProcessor();

        } else if (generationModel.getAnnotationEnum() == AnnotationEnum.AUTO_VALUE_GSON) {
            return Injector.getAppComponent().newAutoValueClassPostProcessor();

        } else {
            return Injector.getAppComponent().newCommonJavaPostProcessor();
        }
    }
}

File: src/main/kotlin/com/robohorse/robopojogenerator/generator/postrocessing/PostProcessorFactory.kt
package com.robohorse.robopojogenerator.generator.postrocessing

import com.robohorse.robopojogenerator.generator.postrocessing.common.AutoValueClassPostProcessor
import com.robohorse.robopojogenerator.generator.postrocessing.common.CommonJavaPostProcessor
import com.robohorse.robopojogenerator.generator.postrocessing.common.KotlinDataClassPostProcessor
import com.robohorse.robopojogenerator.models.GenerationModel
import com.robohorse.robopojogenerator.view.FrameworkVW.AutoValue

internal class PostProcessorFactory(
    private val kotlinDataClassPostProcessor: KotlinDataClassPostProcessor,
    private val autoValueClassPostProcessor: AutoValueClassPostProcessor,
    private val commonJavaPostProcessor: CommonJavaPostProcessor
) {
    fun createPostProcessor(
        generationModel: GenerationModel
    ): BasePostProcessor = with(generationModel) {
        if (useKotlin) {
            kotlinDataClassPostProcessor
        } else if (annotationEnum is AutoValue) {
            autoValueClassPostProcessor
        } else {
            commonJavaPostProcessor
        }
    }
}
